### DISCUSSION TITLE
why are we able to use 'if' inside a 'pmap' train_step?

### DISCUSSION THREAD

**Discussion by yiyixuxu** [google/flax#2472] [Q&A] [answered]
I saw this pattern in the imagenet example - can anyone help me understand why are we able to use `if` inside a `pmap` here? the argument `state` here is a flax `TrainState` type

I really appreciate your help! 

YiYi

```python
@jax.pmap
def train_step(state,...):
   if state.dynamic_scale:
       ....
 
train_step(state,...)

```

https://github.com/google/flax/blob/main/examples/imagenet/train.py#L128

**Comment by andsteing** [comment, marked as answer]
You can use Python control logic inside a `jit`/`pmap`, as long as that control flow is not *value dependent*.

In this case, the if statement is about the value of `state.dynamic_scale` which can be `None`, if no dynamic scaling is used, or am instance of `dynamic_scale_lib.DynamicScale`, if dynamic scaling is used. In either case, the value will be set from the config (see lines 227-230 in code snippet below), and is not changed during the training. In other words, `jit`/`pmap` will produce a single compiled version of the code.

https://github.com/google/flax/blob/6fd24460e88c833fa481228582937c692e9c182c/examples/imagenet/train.py#L222-L244

**Comment by yiyixuxu** [comment]
ohhh thank you! i got it now. it is not actually a `bool` type
